File: dash_apps/models/trip.py
from sqlalchemy import Column, String, Integer, DateTime, Boolean, Float, ForeignKey
from sqlalchemy.orm import relationship
from dash_apps.core.database import Base

class Trip(Base):
    __tablename__ = "trips"

    trip_id = Column(String, primary_key=True)
    created_at = Column(DateTime(timezone=True), nullable=True)
    updated_at = Column(DateTime(timezone=True), nullable=True)
    departure_date = Column(DateTime(timezone=True), nullable=True)
    departure_name = Column(String, nullable=True)
    departure_schedule = Column(DateTime(timezone=True), nullable=True)
    destination_name = Column(String, nullable=True)
    seats_available = Column(Integer, nullable=True)
    seats_booked = Column(Integer, nullable=True)
    seats_published = Column(Integer, nullable=True)
    passenger_price = Column(Integer, nullable=True)
    driver_price = Column(Integer, nullable=True)
    distance = Column(Float, nullable=True)
    precision = Column(String, nullable=True)
    polyline = Column(String, nullable=True)
    driver_id = Column(String, ForeignKey("users.uid"), nullable=True)
    departure_latitude = Column(Float, nullable=True)
    departure_longitude = Column(Float, nullable=True)
    destination_latitude = Column(Float, nullable=True)
    destination_longitude = Column(Float, nullable=True)
    status = Column(String, nullable=True)
    departure_description = Column(String, nullable=True)
    destination_description = Column(String, nullable=True)

    def to_dict(self):
        return {
            "trip_id": self.trip_id,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "departure_date": self.departure_date,
            "departure_name": self.departure_name,
            "departure_schedule": self.departure_schedule,
            "destination_name": self.destination_name,
            "seats_available": self.seats_available,
            "seats_booked": self.seats_booked,
            "seats_published": self.seats_published,
            "passenger_price": self.passenger_price,
            "driver_price": self.driver_price,
            "distance": self.distance,
            "precision": self.precision,
            "polyline": self.polyline,
            "driver_id": self.driver_id,
            "departure_latitude": self.departure_latitude,
            "departure_longitude": self.departure_longitude,
            "destination_latitude": self.destination_latitude,
            "destination_longitude": self.destination_longitude,
            "status": self.status,
            "departure_description": self.departure_description,
            "destination_description": self.destination_description,
        }

File: frontend/src/types/trip.ts
// Type complet depuis Supabase
export interface TripRow {
  trip_id: string;
  driver_id: string | null;
  departure_name: string | null;
  departure_description: string | null;
  departure_latitude: number | null;
  departure_longitude: number | null;
  destination_name: string | null;
  destination_description: string | null;
  destination_latitude: number | null;
  destination_longitude: number | null;
  departure_date: string | null;
  departure_schedule: string | null;
  distance: number | null;
  polyline: string | null;
  seats_published: number | null;
  seats_available: number | null;
  seats_booked: number | null;
  passenger_price: number | null;
  driver_price: number | null;
  status: string | null;
  auto_confirmation: boolean | null;
  precision: string | null;
  created_at: string;
  updated_at: string | null;
}

// Type pour la liste (colonnes minimales)
export interface TripListItem {
  trip_id: string;
  departure_name: string | null;
  destination_name: string | null;
  departure_schedule: string | null;
  distance: number | null;
  seats_available: number | null;
  seats_published: number | null;
  passenger_price: number | null;
  status: string | null;
  driver_id: string | null;
}

// Type pour les détails avec infos conducteur
export interface TripDetail extends TripListItem {
  departure_description: string | null;
  departure_latitude: number | null;
  departure_longitude: number | null;
  destination_description: string | null;
  destination_latitude: number | null;
  destination_longitude: number | null;
  polyline: string | null;
  seats_booked: number | null;
  driver_price: number | null;
  auto_confirmation: boolean | null;
  created_at: string | null;
  // Driver info (joined)
  driver_name: string | null;
  driver_photo: string | null;
  driver_phone: string | null;
  driver_rating: number | null;
  driver_rating_count: number | null;
  driver_verified: boolean | null;
  // Passengers info (joined from bookings)
  passengers: Array<{
    uid: string;
    display_name: string | null;
    photo_url: string | null;
  }>;
}

// Type pour les stats dashboard
export interface TripStats {
  total_trips: number;
  active_trips: number;
  completed_trips: number;
  total_distance: number;
  total_seats_booked: number;
}

// Statuts possibles
export type TripStatus = "PENDING" | "ACTIVE" | "COMPLETED" | "CANCELLED" | "ARCHIVED";

// Type pour la carte (page Map)
export interface TripMapItem {
  trip_id: string;
  departure_name: string | null;
  destination_name: string | null;
  departure_latitude: number | null;
  departure_longitude: number | null;
  destination_latitude: number | null;
  destination_longitude: number | null;
  polyline: string;
  status: string | null;
  departure_schedule: string | null;
  passenger_price: number | null;
  seats_available: number | null;
  seats_published: number | null;
  distance: number | null;
  driver: {
    uid: string;
    display_name: string | null;
    photo_url: string | null;
    rating: number | null;
  } | null;
  passengers: Array<{
    uid: string;
    display_name: string | null;
    photo_url: string | null;
  }>;
}

// Filtres pour la carte
export interface MapFilters {
  status: TripStatus | "ALL";
  dateFrom: string | null;
  dateTo: string | null;
  driverId: string | null;
}

// Type legacy pour compatibilité avec mock data
export interface Trip {
  trip_id: string;
  departure_city: string;
  departure_address: string;
  departure_lat: number;
  departure_lon: number;
  destination_city: string;
  destination_address: string;
  destination_lat: number;
  destination_lon: number;
  trip_distance: number;
  departure_schedule: string;
  price_per_seat: number;
  available_seats: number;
  total_seats: number;
  passengers: string[];
  driver_id: string;
  driver_name: string;
  status: string;
  trip_polyline?: string;
  created_at: string;
  viator_income?: number;
}

// Convertir TripDetail vers Trip (legacy)
export function toTrip(detail: TripDetail): Trip {
  return {
    trip_id: detail.trip_id,
    departure_city: detail.departure_name || "",
    departure_address: detail.departure_description || "",
    departure_lat: detail.departure_latitude || 0,
    departure_lon: detail.departure_longitude || 0,
    destination_city: detail.destination_name || "",
    destination_address: detail.destination_description || "",
    destination_lat: detail.destination_latitude || 0,
    destination_lon: detail.destination_longitude || 0,
    trip_distance: detail.distance || 0,
    departure_schedule: detail.departure_schedule || "",
    price_per_seat: detail.passenger_price || 0,
    available_seats: detail.seats_available || 0,
    total_seats: detail.seats_published || 0,
    passengers: detail.passengers ? detail.passengers.map((p) => p.uid) : [], // Utiliser les vrais IDs passagers
    driver_id: detail.driver_id || "",
    driver_name: detail.driver_name || "",
    status: detail.status || "",
    trip_polyline: detail.polyline || undefined,
    created_at: detail.created_at || "",
    viator_income: detail.driver_price || 0,
  };
}
